File: rs/anda_cognitive_nexus/src/trust.rs
//! Explicit contextual trust and atomic provenance. Calibration algorithms and
//! automatic application policies remain with the trusted Brain host.
use crate::{
    governance::{EffectiveAuthority, Permission, ResourceContext},
    nexus::Session,
    schema::{Intent, SymbolKind, contracts::digest},
    store::Element,
    tx::Transaction,
};
use anda_kip::{Json, KipError, cognitive::ArtifactPin};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::{BTreeMap, BTreeSet};

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct ContextualTrustRule {
    pub id: String,
    pub actor_ref: String,
    pub predicate_ref: Option<String>,
    pub context_ref: Option<String>,
    pub weight: f64,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct TrustConfiguration {
    pub weights: BTreeMap<String, f64>,
    pub default_weight: f64,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub rules: Vec<ContextualTrustRule>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct TrustCalibrationProposal {
    pub format: String,
    pub space_id: String,
    pub expected_version: u64,
    pub configuration: TrustConfiguration,
    pub method: ArtifactPin,
    pub evidence_refs: Vec<String>,
    pub uncertainty: Json,
}

fn invalid(message: &str) -> KipError {
    KipError::constraint_violation(message)
}

impl TrustConfiguration {
    pub fn validate(&self) -> Result<(), KipError> {
        if self.weights.len() > 1024
            || self.rules.len() > 128
            || std::iter::once(&self.default_weight)
                .chain(self.weights.values())
                .chain(self.rules.iter().map(|r| &r.weight))
                .any(|v| !v.is_finite() || !(0.0..=1.0).contains(v))
        {
            return Err(invalid("trust weights must be bounded in [0,1]"));
        }
        let mut ids = BTreeSet::new();
        let mut selectors = BTreeSet::new();
        for rule in &self.rules {
            if rule.id.is_empty()
                || rule.id.len() > 128
                || !ids.insert(&rule.id)
                || (rule.predicate_ref.is_none() && rule.context_ref.is_none())
                || !selectors.insert((&rule.actor_ref, &rule.predicate_ref, &rule.context_ref))
            {
                return Err(invalid(
                    "contextual trust needs unique IDs/selectors and an explicit context or predicate",
                ));
            }
        }
        Ok(())
    }

    /// More specific rules win. Conflicting equal-specificity matches fail
    /// closed; rule ordering never chooses an implicit winner.
    pub fn weight(
        &self,
        actor: &str,
        predicate: &str,
        contexts: &[String],
    ) -> Result<f64, KipError> {
        weight(
            &self.rules,
            self.weights
                .get(actor)
                .copied()
                .unwrap_or(self.default_weight),
            actor,
            predicate,
            contexts,
        )
    }
}

pub(crate) fn weight(
    rules: &[ContextualTrustRule],
    fallback: f64,
    actor: &str,
    predicate: &str,
    contexts: &[String],
) -> Result<f64, KipError> {
    let mut selected: Option<(usize, f64)> = None;
    let mut ambiguous = false;
    for rule in rules {
        if rule.actor_ref != actor
            || rule.predicate_ref.as_ref().is_some_and(|p| p != predicate)
            || rule
                .context_ref
                .as_ref()
                .is_some_and(|c| !contexts.contains(c))
        {
            continue;
        }
        let specificity =
            usize::from(rule.predicate_ref.is_some()) + usize::from(rule.context_ref.is_some());
        match selected {
            Some((old, value)) if specificity == old => {
                ambiguous |= value != rule.weight;
            }
            Some((old, _)) if old > specificity => {}
            _ => {
                selected = Some((specificity, rule.weight));
                ambiguous = false;
            }
        }
    }
    if ambiguous {
        return Err(invalid(
            "ambiguous contextual trust; use an explicit combined scope",
        ));
    }
    Ok(selected.map(|(_, v)| v).unwrap_or(fallback))
}

async fn validate_references(
    session: &Session,
    authority: &EffectiveAuthority,
    space: &str,
    configuration: &TrustConfiguration,
) -> Result<(), KipError> {
    configuration.validate()?;
    let env = session.nexus.store.schema_environment(space).await?;
    for rule in &configuration.rules {
        for reference in std::iter::once(&rule.actor_ref).chain(rule.context_ref.iter()) {
            let id = crate::ElementId::parse_kind(reference, anda_kip::ElementKind::Concept)?;
            let row = session.nexus.store.get_element(id).await?;
            if row.space() != space || id.to_string() != *reference {
                return Err(KipError::not_found_or_not_visible(
                    "trust context unavailable",
                ));
            }
            authority
                .authorize(
                    Permission::Read,
                    &ResourceContext::of_element(&row),
                    &session.auth,
                )
                .into_result()?;
        }
        if let Some(predicate) = &rule.predicate_ref
            && env
                .resolve_symbol(SymbolKind::PredicateType, predicate, Intent::Read)?
                .to_string()
                != *predicate
        {
            return Err(invalid(
                "trust predicates must use their exact schema references",
            ));
        }
    }
    Ok(())
}

impl Session {
    /// Explicit replacement of global and contextual settings; no algorithm is
    /// run and no cognitive text is interpreted as a control-plane instruction.
    pub async fn set_contextual_trust(
        &self,
        space: &str,
        expected: u64,
        configuration: TrustConfiguration,
    ) -> Result<Json, KipError> {
        self.governed(space, Permission::ManageTrust, async || {
            let authority = self.effective_authority(space).await?;
            validate_references(self, &authority, space, &configuration).await?;
            let row = self
                .nexus
                .store
                .publish_control(
                    space,
                    "trust",
                    "trust",
                    expected,
                    json!(configuration),
                    json!({"principal_id":self.auth.principal_id}),
                )
                .await?;
            Ok(json!({"version":row.version,"trust":row.value}))
        })
        .await
    }

    /// Commit the exact proposed configuration, provenance and Governance audit
    /// in one redo plan. ManageTrust remains necessary; no Create permission is
    /// implied or needed merely to record the native Governance audit.
    pub async fn apply_trust_calibration(
        &self,
        space: &str,
        expected: u64,
        proposal: ArtifactPin,
        operation_key: &str,
    ) -> Result<Json, KipError> {
        if operation_key.is_empty() || operation_key.len() > 256 {
            return Err(invalid("bounded trust operation key required"));
        }
        self.governed(space, Permission::ManageTrust, async || {
            let store = &self.nexus.store;
            let authority = self.effective_authority(space).await?;
            let (content, sources) = store
                .authorized_artifact(space, &proposal.artifact_ref, &authority, &self.auth)
                .await?;
            if digest(&content)? != proposal.content_digest {
                return Err(invalid("proposal digest mismatch"));
            }
            let proposed: TrustCalibrationProposal =
                serde_json::from_value(content).map_err(|_| invalid("invalid trust proposal"))?;
            if proposed.format != "nexus:trust-calibration-v1"
                || proposed.space_id != space
                || proposed.expected_version != expected
                || proposed.evidence_refs.is_empty()
                || proposed.evidence_refs.len() > 128
                || proposed
                    .uncertainty
                    .as_object()
                    .is_none_or(|v| v.is_empty())
            {
                return Err(invalid(
                    "trust calibration needs scope, evidence and explicit uncertainty",
                ));
            }
            validate_references(self, &authority, space, &proposed.configuration).await?;
            let (method, _) = store
                .authorized_artifact(space, &proposed.method.artifact_ref, &authority, &self.auth)
                .await?;
            if digest(&method)? != proposed.method.content_digest {
                return Err(invalid("calibration method digest mismatch"));
            }
            let mut evidence = BTreeSet::new();
            for reference in &proposed.evidence_refs {
                if !evidence.insert(reference) || !sources.contains(reference) {
                    return Err(invalid(
                        "proposal must inherit every independent evidence material reference",
                    ));
                }
                let row = store.get_element(reference.parse()?).await?;
                let Element::Evidence(record) = &row else {
                    return Err(invalid("calibration evidence must be Evidence"));
                };
                if row.space() != space
                    || record.status == "corrected"
                    || row.state() != crate::store::rows::state::ACTIVE
                {
                    return Err(invalid("calibration evidence is no longer eligible"));
                }
                authority
                    .authorize(
                        Permission::Read,
                        &ResourceContext::of_element(&row),
                        &self.auth,
                    )
                    .into_result()?;
            }
            let identity = json!({"operation":"apply_trust_calibration","key":operation_key});
            let key = crate::attention::request_key(&self.auth.principal_id, &identity)?;
            let request_digest = digest(&json!({"expected":expected,"proposal":proposal}))?;
            if let Some(result) = crate::attention::replay(store, space, &key, &request_digest).await? {
                return Ok(result);
            }
            let mut tx = Transaction::begin(
                store,
                space,
                json!({"principal_id":self.auth.principal_id}),
                false,
                authority,
                (*self.auth).clone(),
            )
            .await?;
            let mut value = json!(proposed.configuration);
            value["calibration"] = json!({"proposal":proposal,"method":proposed.method,"evidence_refs":proposed.evidence_refs});
            crate::attention::stage_control(store, &mut tx, "trust", expected, "trust", value).await?;
            tx.defer_governance_audit(crate::governance::store::MutationEntry {
                        at:tx.cx.at.clone(),
                        space_id:space.into(),principal_id:self.auth.principal_id.clone(),operation:"apply_trust_calibration",resource:"trust".into(),
                        record:json!({"proposal":proposal,"method":proposed.method,"before_version":expected,"after_version":expected+1}),
                    });
            crate::attention::commit(store,tx,key,request_digest,json!({"version":expected+1,"proposal":proposal,"audit_operation":"apply_trust_calibration"})).await
        })
        .await
    }
}

File: ts/kip-do/src/trust.ts
/** Explicit trust settings and governed calibration provenance; algorithms live in the host. */
import type { Session } from './nexus.js'
import type { Json, JsonMap } from './json.js'
import type { ArtifactPin } from './cognitive.js'
import { isJsonMap } from './json.js'
import { authorizedArtifact, publishControl } from './control.js'
import { parseElementId, formatElementId } from './id.js'
import { formatSymbolRef } from './schema/symbol.js'
import { digest } from './schema/contracts.js'
import { errors } from './errors.js'
import {
  bytes,
  json,
  exact,
  safe,
  invalid,
  permit,
  requestKey,
  transaction,
  stageControl,
  replay,
  commit,
} from './attention/common.js'

export interface ContextualTrustRule {
  id: string
  actor_ref: string
  predicate_ref: string | null
  context_ref: string | null
  weight: number
}
export interface TrustConfiguration {
  weights: Record<string, number>
  default_weight: number
  rules?: ContextualTrustRule[]
}
export interface TrustCalibrationProposal {
  format: string
  space_id: string
  expected_version: number
  configuration: TrustConfiguration
  method: ArtifactPin
  evidence_refs: string[]
  uncertainty: Json
}
export function validateTrust(configuration: TrustConfiguration): void {
  exact(configuration, ['weights', 'default_weight', 'rules'])
  if (
    !isJsonMap(json(configuration.weights)) ||
    (configuration.rules !== undefined && !Array.isArray(configuration.rules))
  )
    invalid('invalid trust configuration')
  const rules = configuration.rules ?? []
  if (
    Object.keys(configuration.weights).length > 1024 ||
    rules.length > 128 ||
    [
      configuration.default_weight,
      ...Object.values(configuration.weights),
      ...rules.map((r) => r.weight),
    ].some(
      (v) => typeof v !== 'number' || !Number.isFinite(v) || v < 0 || v > 1,
    )
  )
    invalid('trust weights must be bounded in [0,1]')
  const ids = new Set<string>(),
    selectors = new Set<string>()
  for (const rule of rules) {
    exact(rule, ['id', 'actor_ref', 'predicate_ref', 'context_ref', 'weight'])
    const selector = JSON.stringify([
      rule.actor_ref,
      rule.predicate_ref ?? null,
      rule.context_ref ?? null,
    ])
    if (
      typeof rule.id !== 'string' ||
      !rule.id ||
      bytes(rule.id) > 128 ||
      ids.has(rule.id) ||
      (!rule.predicate_ref && !rule.context_ref) ||
      selectors.has(selector)
    )
      invalid(
        'contextual trust needs unique IDs/selectors and an explicit context or predicate',
      )
    ids.add(rule.id)
    selectors.add(selector)
  }
}
export function trustWeight(
  rules: ContextualTrustRule[],
  fallback: number,
  actor: string,
  predicate: string,
  contexts: string[],
): number {
  let selected: { specificity: number; weight: number } | null = null,
    ambiguous = false
  for (const rule of rules) {
    if (
      rule.actor_ref !== actor ||
      (rule.predicate_ref != null && rule.predicate_ref !== predicate) ||
      (rule.context_ref != null && !contexts.includes(rule.context_ref))
    )
      continue
    const specificity =
      Number(rule.predicate_ref != null) + Number(rule.context_ref != null)
    if (selected && specificity === selected.specificity)
      ambiguous ||= selected.weight !== rule.weight
    else if (!selected || specificity > selected.specificity) {
      selected = { specificity, weight: rule.weight }
      ambiguous = false
    }
  }
  if (ambiguous)
    invalid('ambiguous contextual trust; use an explicit combined scope')
  return selected?.weight ?? fallback
}
function validateReferences(
  s: Session,
  space: string,
  config: TrustConfiguration,
): void {
  validateTrust(config)
  for (const rule of config.rules ?? []) {
    for (const ref of [
      rule.actor_ref,
      ...(rule.context_ref != null ? [rule.context_ref] : []),
    ]) {
      const id = parseElementId(ref),
        row = s.nexus.store.load(id)
      if (
        id.kind !== 'Concept' ||
        formatElementId(id) !== ref ||
        !row ||
        row.row.space !== space
      )
        throw errors.notFoundOrNotVisible('trust context unavailable')
      permit(s, space, 'read', row)
    }
    if (
      rule.predicate_ref != null &&
      formatSymbolRef(
        s.nexus
          .environment(space)
          .resolveSymbol('PredicateType', rule.predicate_ref, 'read'),
      ) !== rule.predicate_ref
    )
      invalid('trust predicates must use their exact schema references')
  }
}
export function setContextualTrust(
  s: Session,
  space: string,
  expected: number,
  configuration: TrustConfiguration,
): JsonMap {
  validateReferences(s, space, configuration)
  const value = {
    weights: configuration.weights,
    default_weight: configuration.default_weight,
    ...(configuration.rules?.length
      ? {
          rules: configuration.rules.map((r) => ({
            ...r,
            predicate_ref: r.predicate_ref ?? null,
            context_ref: r.context_ref ?? null,
          })),
        }
      : {}),
  }
  const row = publishControl(
    s.nexus.store,
    space,
    'trust',
    'trust',
    expected,
    json(value),
    { principal_id: s.auth.principal_id },
  )
  return { version: row.version, trust: row.value }
}
export function applyTrustCalibration(
  s: Session,
  space: string,
  expected: number,
  proposal: ArtifactPin,
  operationKey: string,
  authorizeWrite: () => void,
): JsonMap {
  if (
    typeof operationKey !== 'string' ||
    !operationKey ||
    bytes(operationKey) > 256 ||
    !safe(expected)
  )
    invalid('bounded trust operation key required')
  const store = s.nexus.store,
    authority = s.effectiveAuthority(space),
    { payload, sources } = authorizedArtifact(
      store,
      space,
      proposal.artifact_ref,
      authority,
      s.auth,
    )
  if (digest(payload) !== proposal.content_digest)
    invalid('proposal digest mismatch')
  exact(payload, [
    'format',
    'space_id',
    'expected_version',
    'configuration',
    'method',
    'evidence_refs',
    'uncertainty',
  ])
  const proposed = payload as unknown as TrustCalibrationProposal
  if (
    proposed.format !== 'nexus:trust-calibration-v1' ||
    proposed.space_id !== space ||
    proposed.expected_version !== expected ||
    !Array.isArray(proposed.evidence_refs) ||
    !proposed.evidence_refs.length ||
    proposed.evidence_refs.length > 128 ||
    !isJsonMap(proposed.uncertainty) ||
    !Object.keys(proposed.uncertainty).length
  )
    invalid('trust calibration needs scope, evidence and explicit uncertainty')
  validateReferences(s, space, proposed.configuration)
  const method = authorizedArtifact(
    store,
    space,
    proposed.method.artifact_ref,
    authority,
    s.auth,
  ).payload
  if (digest(method) !== proposed.method.content_digest)
    invalid('calibration method digest mismatch')
  const evidence = new Set<string>()
  for (const ref of proposed.evidence_refs) {
    if (evidence.has(ref) || !sources.includes(ref))
      invalid(
        'proposal must inherit every independent evidence material reference',
      )
    evidence.add(ref)
    const row = store.load(parseElementId(ref))
    if (
      !row ||
      row.kind !== 'Evidence' ||
      row.row.space !== space ||
      row.row.status === 'corrected' ||
      row.row.state !== 'active'
    )
      invalid('calibration evidence is no longer eligible')
    permit(s, space, 'read', row)
  }
  const key = requestKey(s, {
      operation: 'apply_trust_calibration',
      key: operationKey,
    }),
    hash = digest(json({ expected, proposal })),
    previous = replay(s, space, key, hash)
  if (previous) return previous
  authorizeWrite()
  const tx = transaction(s, space)
  stageControl(tx, 'trust', expected, 'trust', {
    ...proposed.configuration,
    calibration: {
      proposal,
      method: proposed.method,
      evidence_refs: proposed.evidence_refs,
    },
  })
  // Both writes are enclosed by the Session's SQLite transaction.
  store.governance.recordMutation({
    at: tx.cx.at,
    space_id: space,
    principal_id: s.auth.principal_id,
    operation: 'apply_trust_calibration',
    resource: 'trust',
    record: json({
      proposal,
      method: proposed.method,
      before_version: expected,
      after_version: expected + 1,
    }),
  })
  return commit(s, tx, key, hash, {
    version: expected + 1,
    proposal: json(proposal),
    audit_operation: 'apply_trust_calibration',
  })
}
